### ISSUE TITLE
False unsubscriptable-object warning

### ISSUE THREAD

**Issue by bje-** [pylint-dev/pylint#9250]
### Bug description

```python
Pylint produces an incorrect unsubscriptable-object warning.


"""A test case."""
import numpy as np

class Foo:
    """Class docstring."""

    csvfilename = None
    csvdata = None

    def __init__(self, filename):
        cls = self.__class__
        assert filename is not None
        if cls.csvfilename != filename:
            cls.csvdata = np.zeros(10)
            cls.csvfilenanme = filename
        self.generation = cls.csvdata[0]
```
```


### Configuration

_No response_

### Command used

```shell
pylint testcase.py
```

### Pylint output

```python
************* Module testcase
testcase.py:16:26: E1136: Value 'cls.csvdata' is unsubscriptable (unsubscriptable-object)
```

### Expected behavior

No warning should be emitted.

### Pylint version

```shell
pylint 3.0.2
astroid 3.0.1
Python 3.10.12 (main, Jun 11 2023, 05:26:28) [GCC 11.4.0]
```

### OS / Environment

Ubuntu 22.04

**Comment by jacobtylerwalls** [comment]
Thanks for the report. pylint/astroid currently doesn't understand dynamic assignment to class attributes, like `cls.csvdata = ...`. We just know that `cls.csvdata` is None from the earlier assignment.

Demo:
```py
>>> import astroid
>>> node = astroid.extract_node("""
... class MyClass:
...     attr = None
...     def __init__(self):
...          attr = 1
... 
... MyClass.attr #@
... """)
>>> node
<Attribute.attr l.7 at 0x1084e1040>
>>> node.inferred()
[<Const.NoneType l.3 at 0x1046bb3b0>]
```

**Comment by jacobtylerwalls** [comment]
Sorry, here's the correct demo:

```py
>>> node = astroid.extract_node("""
... class MyClass:
...     attr = None
...     def __init__(self):
...         cls = self.__class__
...         cls.attr = 1
...         cls.attr #@
... """)
>>> node.inferred()
[<Const.NoneType l.3 at 0x1084e2ea0>]
```

**Comment by jacobtylerwalls** [comment]
Actually, the warning is correct if the equality `if cls.csvfilename != filename` fails, but I take your point.
